src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -603,6 +604,132 @@ case class MapEntries(child: MatrixIR, newEntries: IR) extends MatrixIR {
   }
 }
 
+case class MapRows(child: MatrixIR, newRow: IR) extends MatrixIR {
+
+  def children: IndexedSeq[BaseIR] = Array(child, newRow)
+
+  def copy(newChildren: IndexedSeq[BaseIR]): MapEntries = {
+    assert(newChildren.length == 2)
+    MapEntries(newChildren(0).asInstanceOf[MatrixIR], newChildren(1).asInstanceOf[IR])
+  }
+
+  val newRVRow = InsertFields(newRow, Seq(MatrixType.entriesIdentifier -> GetField(Ref("va"), MatrixType.entriesIdentifier)))
```

Comment:
This immediately suggests an optimization rule that collapses nested InsertFields (and a MakeStruct in an InsertField).  For another PR.